modules/wallet.go

```diff
@@ -22,68 +24,102 @@ type WalletInfo struct {
+// unsigned form) by calling 'ViewTransaction'.
+//
+// Transactions are kept with their parents. A parent is any unconfirmed
+// transaction that is required for the child to be valid.
 type TransactionBuilder interface {
-	// RegisterTransaction creates a transaction out of an existing transaction
-	// which can be modified by the wallet, returning an id that can be used to
-	// reference the transaction.
-	RegisterTransaction(types.Transaction) (id string, err error)
-
-	// FundTransaction will add `amount` to a transaction's inputs. The funded
-	// transaction is returned with an error.
-	FundTransaction(id string, amount types.Currency) (types.Transaction, error)
-
-	// AddMinerFee adds a single miner fee of value `fee`. The transaction is
-	// returned, along with the index that the added fee ended up at.
-	AddMinerFee(id string, fee types.Currency) (types.Transaction, uint64, error)
-
-	// AddSiacoinInput adds a siacoin input to the transaction. When
-	// 'SignTransaction' gets called, this input will be left unsigned. The
-	// updated transaction is returned along with the index of the new siacoin
-	// input within the transaction.
-	AddSiacoinInput(id string, input types.SiacoinInput) (types.Transaction, uint64, error)
-
-	// AddSiacoinOutput adds an output to a transaction. It returns the transaction
-	// with index of the output that got added.
-	AddSiacoinOutput(id string, output types.SiacoinOutput) (types.Transaction, uint64, error)
-
-	// AddFileContract adds a file contract to a transaction, returning the
-	// transaction and the index that the file contract was put at.
-	AddFileContract(id string, fc types.FileContract) (types.Transaction, uint64, error)
-
-	// AddStorageProof adds a storage proof to a transaction, returning the
-	// transaction and the index that the storage proof was put at.
-	AddStorageProof(id string, sp types.StorageProof) (types.Transaction, uint64, error)
-
-	// AddSiafundInput adds a siacoin input to the transaction. When
-	// 'SignTransaction' gets called, this input will be left unsigned. The
-	// updated transaction is returned along with the index of the new siacoin
-	// input within the transaction.
-	AddSiafundInput(id string, input types.SiafundInput) (types.Transaction, uint64, error)
-
-	// AddSiafundOutput adds an output to a transaction. It returns the transaction
-	// with index of the output that got added.
-	AddSiafundOutput(id string, output types.SiafundOutput) (types.Transaction, uint64, error)
+	// RegisterTransaction takes a transaction and its parents returns an id
+	// that can be used to modify the transaction. The most typical call is
+	// 'RegisterTransaction(types.Transaction{}, nil)', which registers a new
+	// transaction that doesn't have any parents. The id that gets returned is
+	// not a types.TransactionID, it is an int and is only useful within the
+	// transaction builder.
+	RegisterTransaction(t types.Transaction, parents []types.Transaction) (id int, err error)
```

Comment:
NewTransaction is a bad name because it's not a transaction, it's a transaction + a bunch of additional data.

I'm not sure why I didn't think about returning a transaction object, the register model is kind of clunky, AND you have to worry about leaks. I think it would be possible to adjust the builder to return an object instead of an id, but that's a bigger change than I want to tackle in this PR.